File: embroidermodder/actions.c
/* Embroidermodder 2.
 * ------------------------------------------------------------
 * Copyright 2021 The Embroidermodder Team
 * Embroidermodder 2 is Open Source Software.
 * See LICENSE.txt for licensing terms.
 * ------------------------------------------------------------
 * This file is for the functions, not the data, of embroidermodder 2.
 */

#include "embroidermodder.h"

#include <string.h>
#include <stdlib.h>

void icon16(void)
{
    debug_message("icon16()");
}

void icon24(void)
{
    debug_message("icon24()");
}

void icon32(void)
{
    debug_message("icon32()");
}

void icon48(void)
{
    debug_message("icon48()");
}

void icon64(void)
{
    debug_message("icon64()");
}

void icon128(void)
{
    debug_message("icon128()");
}

void newFile(void)
{
    debug_message("newFile()");
}

void openFile(void)
{
    debug_message("openFile()");
}

void saveFile(void)
{
    debug_message("saveFile()");
}

void main_print(void)
{
    debug_message("print()");
}

void main_exit(void)
{
    debug_message("main_exit()");
    exit(0);
}

void saveAsFile(void)
{

}

void whatsthisContextHelp(void)
{

}

void makeLayerCurrent(void)
{

}

void layerSelector(void)
{

}

void main_about(void)
{
    /*TODO: QTabWidget for about dialog*/
    /*QApplication::setOverrideCursor(Qt::ArrowCursor);
    debug_message("about()");
    QString appDir = qApp->applicationDirPath();
    QString title = "About Embroidermodder 2";

    QDialog dialog(_mainWin);
    ImageWidget img(appDir + "/images/logo-small.png");
    QLabel text(QString("Embroidermodder ") + QString(_appVer_) + "\n\n" +
                          _mainWin->tr("http://embroidermodder.org") +
                          "\n\n" +
                          _mainWin->tr("Available Platforms: GNU/Linux, Windows, Mac OSX, Raspberry Pi") +
                          "\n\n" +
                          _mainWin->tr("Embroidery formats by Josh Varga.") +
                          "\n" +
                          _mainWin->tr("User Interface by Jonathan Greig and Robin Swift.") +
                          "\n\n" +
                          _mainWin->tr("Free under the zlib/libpng license.")
                          #if defined(BUILD_GIT_HASH)
                          + "\n\n" +
                          _mainWin->tr("Build Hash: ") + qPrintable(BUILD_GIT_HASH)
                          #endif
                          );
    text.setWordWrap(1);

    QDialogButtonBox buttonbox(Qt::Horizontal, &dialog);
    QPushButton button(&dialog);
    button.setText("Oh, Yeah!");
    buttonbox.addButton(&button, QDialogButtonBox::AcceptRole);
    buttonbox.setCenterButtons(1);
    _mainWin->connect(&buttonbox, SIGNAL(accepted()), &dialog, SLOT(accept()));

    QVBoxLayout layout;
    layout.setAlignment(Qt::AlignCenter);
    layout.addWidget(&img);
    layout.addWidget(&text);
    layout.addWidget(&buttonbox);

    dialog.setWindowTitle(title);
    dialog.setMinimumWidth(img.minimumWidth()+30);
    dialog.setMinimumHeight(img.minimumHeight()+50);
    dialog.setLayout(&layout);
    dialog.exec();
    QApplication::restoreOverrideCursor();
    */
}

void main_help(void)
{

}

void settingsDialog(void)
{

}

void designDetails(void)
{

}

void main_cut(void)
{
    debug_message("cut()");
    /*
    View* gview = _mainWin->activeView();
    if (gview) {
        gview->cut();
    }
    */
}

void main_copy(void)
{
    debug_message("copy()");
    /*
    View* gview = _mainWin->activeView();
    if (gview) {
        gview->copy();
    }
    */
}

void main_paste(void)
{
    debug_message("main_paste()");
    /*
    View* gview = _mainWin->activeView();
    if (gview) {
        gview->paste();
    }
    */
}


void main_redo(void)
{
    debug_message("copy()");
    /*
    View* gview = _mainWin->activeView();
    if (gview) {
        gview->copy();
    }
    */
}

void main_undo(void)
{
    debug_message("main_paste()");
    /*
    View* gview = _mainWin->activeView();
    if (gview) {
        gview->paste();
    }
    */
}

void tipOfTheDay(void)
{

}

void changelog(void)
{
    debug_message("changelog()");

    /* display in a custom widget instead */
    /*
    QUrl changelogURL("help/changelog.html");
    QDesktopServices::openUrl(changelogURL);
    */
}

void showAllLayers(void)
{

}

void freezeAllLayers(void)
{

}

void thawAllLayers(void)
{

}

void lockAllLayers(void)
{

}

void unlockAllLayers(void)
{

}

void hideAllLayers(void)
{

}

void lineWeightSelector(void)
{

}

void lineTypeSelector(void)
{

}

void colorSelector(void)
{

}

void windowClose(void)
{

}

void windowTile(void)
{

}

void windowCloseAll(void)
{

}

void windowCascade(void)
{

}

void windowNext(void)
{

}

void windowPrevious(void)
{

}

void textItalic(void)
{
    settings.text_style.italic = !settings.text_style.italic;
}

void textBold(void)
{
    settings.text_style.bold = !settings.text_style.bold;
}

void textStrikeout(void)
{
    settings.text_style.strikeout = !settings.text_style.strikeout;
}

void textUnderline(void)
{
    settings.text_style.underline = !settings.text_style.underline;
}

void textOverline(void)
{
    settings.text_style.overline = !settings.text_style.overline;
}


void makeLayerActive(void)
{
    debug_message("makeLayerActive()");
    debug_message("Implement makeLayerActive.");
}

void layerManager(void)
{
    debug_message("layerManager()");
    debug_message("Implement layerManager.");
    /*LayerManager layman( _mainWin,  _mainWin);
    layman.exec();
    */
}

void layerPrevious(void)
{
    debug_message("layerPrevious()");
    debug_message("Implement layerPrevious.");
}

void zoomRealtime(void)
{
    debug_message("zoomRealtime()");
    debug_message("Implement zoomRealtime.");
}

void zoomPrevious(void)
{
    debug_message("zoomPrevious()");
    debug_message("Implement zoomPrevious.");
}

void zoomWindow(void)
{
    debug_message("zoomWindow()");
    /*View* gview = _mainWin->activeView();
    if (gview) {
        gview->zoomWindow();
    }
    */
}

void zoomDynamic(void)
{
    debug_message("zoomDynamic()");
    debug_message("Implement zoomDynamic.");
}

void zoomScale(void)
{
    debug_message("zoomScale()");
    debug_message("Implement zoomScale.");
}

void zoomCenter(void)
{
    debug_message("zoomCenter()");
    debug_message("Implement zoomCenter.");
}

void zoomIn(void)
{
    debug_message("zoomIn()");
}

void zoomOut(void)
{
    debug_message("zoomOut()");
}

void zoomSelected(void)
{
    debug_message("zoomSelected()");
}

void zoomAll(void)
{
    debug_message("zoomAll()");
    debug_message("Implement zoomAll.");
}

void zoomExtents(void)
{
    debug_message("zoomExtents()");
}

void panrealtime(void)
{
    debug_message("panrealtime()");
}

void panpoint(void)
{
    debug_message("panpoint()");
}

void panLeft(void)
{
    debug_message("panLeft()");
}

void panRight(void)
{
    debug_message("panRight()");
}

void panUp(void)
{
    debug_message("panUp()");
}

void panDown(void)
{
    debug_message("panDown()");
}

void dayVision(void)
{
    #if 0
    View* gview = _mainWin->activeView();
    if (gview) {
        gview->setBackgroundColor(qRgb(255,255,255)); /*TODO: Make day vision color settings.*/
        gview->setCrossHairColor(qRgb(0,0,0));        /*TODO: Make day vision color settings.*/
        gview->setGridColor(qRgb(0,0,0));             /*TODO: Make day vision color settings.*/
    }
    #endif
}

void nightVision(void)
{
    #if 0
    View* gview = _mainWin->activeView();
    if (gview) {
        gview->setBackgroundColor(qRgb(0,0,0)); /* TODO: Make night vision color settings. */
        gview->setCrossHairColor(qRgb(255,255,255)); /*TODO: Make night vision color settings.*/
        gview->setGridColor(qRgb(255,255,255));      /*TODO: Make night vision color settings.*/
    }
    #endif
}

void doNothing(void)
{
    /* This function intentionally does nothing. */
    debug_message("doNothing()");
}


void actuator(char *call)
{
    int id;
    undo_history_position++;
    /* an action has been taken, we are at the current head of the stack */
    undo_history_length = undo_history_position;
    strcpy(undo_history[undo_history_position], call);
    id = call[0];
    if (id < 0) {
        id += 256;
    }
    if (id < N_ACTIONS) {
        action_list[id].function();
    }
}

/* --------------------------------------------------------------- */

void get_n_ints(const char *command, int *out, int n);
void get_n_floats(const char *command, float *out, int n);

/* This is similar to using an svg path, we can blend these systems
 * later. */
 #if 0
QPixmap *draw_pixmap(const char *description)
{
    char *ptr;
    int int_buffer[4];
    QPixmap *icon;
    QPainter *painter;
    QPen pen;
    get_n_ints(description, int_buffer, 2);
    icon = new QPixmap(int_buffer[0], int_buffer[1]);
    painter = new QPainter(icon);
    pen.setWidth(10);
    for (ptr=(char*)description; *ptr; ptr++) {
        /* Other functions we can use are eraseRect, drawArc etc. https://doc.qt.io/qt-5/qpainter.html */
        if (strncmp(ptr, "rect", 4)==0) {
            pen.setColor(QColor(QRgb(0x000000)));
            painter->setPen(pen);
            get_n_ints(ptr+5, int_buffer, 4);
            painter->fillRect(int_buffer[0], int_buffer[1],
                int_buffer[2], int_buffer[3], Qt::SolidPattern); 
        }
    }
    return icon;
}

QIcon loadIcon(const char **icon)
{
    /* so we can experiment with different icon generation methods */
    if (icon[0][0] == 'C') {
        return QIcon(*draw_pixmap(icon[0]+2));
    }
    return QIcon(QPixmap(icon));
}

#endif
void get_n_ints(const char *command, int *out, int n)
{
    int i;
    char modifyable[100];
    strcpy(modifyable, command);
    char *rest = (char*)modifyable;
    for (i=0; i<n; i++) {
        char *tok = strtok_r(rest, " ", &rest);
        out[i] = atoi(tok);
    }
}

void get_n_floats(const char *command, float *out, int n)
{
    int i;
    char modifyable[100];
    strcpy(modifyable, command);
    char *rest = (char*)modifyable;
    for (i=0; i<n; i++) {
        char *tok = strtok_r(rest, " ", &rest);
        out[i] = atof(tok);
    }
}

#if 0
void add_to_path(QPainterPath *path, const char *command, float pos[2], float scale[2])
{
    int j;
    float out[10];
    for (j=0; j<strlen(command); j++) {
        switch (command[j]) {
        case 'M':
            get_n_floats(command+j+2, out, 2);
            path->moveTo(pos[0]+out[0]*scale[0], pos[1]+out[1]*scale[1]);
            break;
        case 'L':
            get_n_floats(command+j+2, out, 2);
            path->lineTo(pos[0]+out[0]*scale[0], pos[1]+out[1]*scale[1]);
            break;
        case 'A':
            get_n_floats(command+j+2, out, 6);
            path->arcTo(pos[0]+out[0]*scale[0], pos[1]+out[1]*scale[1],
                        out[2], out[3], out[4], out[5]);
            break;
        case 'a':
            get_n_floats(command+j+2, out, 5);
            path->arcMoveTo(pos[0]+out[0]*scale[0], pos[1]+out[1]*scale[1],
                        out[2]*scale[0], out[3]*scale[1],
                        out[4]);
            break;
        case 'E':
            get_n_floats(command+j+2, out, 4);
            path->addEllipse(
                QPointF(pos[0]+out[0]*scale[0],  pos[1]+out[1]*scale[1]),
                out[2]*scale[0], out[3]*scale[1]);
            break;
        case 'Z':
            path->closeSubpath();
            break;
        default:
            break;
        }
    }
}

void add_list_to_path(QPainterPath *path, const char *commands[], float pos[2], float scale[2])
{
    for (int i=0; origin_string[i][0]; i++) {
        add_to_path(path, origin_string[i], pos, scale);
    }
}

/*NOTE: This function should be used to interpret various object types and save them as polylines for stitchOnly formats.*/
void toPolyline(EmbPattern* pattern, const QPointF& objPos, const QPainterPath& objPath, const QString& layer, const QColor& color, const QString& lineType, const QString& lineWeight)
{
    float startX = objPos.x();
    float startY = objPos.y();
    EmbArray* pointList = embArray_create(EMB_POINT);
    QPainterPath::Element element;
    for(int i = 0; i < objPath.elementCount(); ++i)
    {
        element = objPath.elementAt(i);
        EmbPointObject a;
        a.point.x = element.x + startX;
        a.point.y = -(element.y + startY);
        embArray_addPoint(pointList, &a);
    }
    EmbPolylineObject* polyObject;
    polyObject = (EmbPolylineObject *) malloc(sizeof(EmbPolylineObject));
    polyObject->pointList = pointList;
    polyObject->color = embColor_make(color.red(), color.green(), color.blue());
    polyObject->lineType = 1; /*TODO: proper lineType*/
    embPattern_addPolylineObjectAbs(pattern, polyObject);
}
#endif

void settingsSnap()
{

}

void settingsGrid()
{

}

void settingsRuler()
{

}

void settingsOrtho()
{

}

void settingsPolar()
{

}

void settingsQSnap()
{

}

void settingsQTrack()
{

}

void settingsLwt()
{

}

void toggleSnap(int on)
{
    debug_message("StatusBarButton toggleSnap()");

}

void toggleGrid(int on)
{
    debug_message("StatusBarButton toggleGrid()");

}

void toggleRuler(int on)
{
    debug_message("StatusBarButton toggleRuler()");

}

void toggleOrtho(int on)
{
    debug_message("StatusBarButton toggleOrtho()");

}

void togglePolar(int on)
{
    debug_message("StatusBarButton togglePolar()");

}

void toggleQSnap(int on)
{
    debug_message("StatusBarButton toggleQSnap()");

}

void toggleQTrack(int on)
{
    debug_message("StatusBarButton toggleQTrack()");

}

void toggleLwt(int on)
{
    debug_message("StatusBarButton toggleLwt()");

}

void enableLwt()
{
    debug_message("StatusBarButton enableLwt()");

}

void disableLwt()
{
    debug_message("StatusBarButton disableLwt()");
}

void enableReal()
{
    debug_message("StatusBarButton enableReal()");
}

void disableReal()
{
    debug_message("StatusBarButton disableReal()");
}

File: embroidermodder/actions.py
#!/usr/bin/env python3

r"""
    Embroidermodder 2.

    ------------------------------------------------------------

    Copyright 2013-2022 The Embroidermodder Team
    Embroidermodder 2 is Open Source Software.
    See LICENSE for licensing terms.

    ------------------------------------------------------------

    To make the undo history easier to manage we use a dict for
    keeping all the action information together.
"""

import tkinter as tk
from PIL import Image, ImageTk
from .data import load_image

debug_mode = True

def debugMessage(s):
    r"""
    Guards against debug messages coming up during normal operation.

    Just change debug_mode to True to activate it. We could have a toggle
    in the program to turn it on during operation for when something starts
    acting weird.
    """
    if debug_mode:
        print(s)


def doNothing():
    r""" This function intentionally does nothing. """
    debugMessage("doNothing()")    


def newFile():
    " Creates a new file using the standard dialog. "
    debugMessage("New File")


def openFile():
    " Opens a file using the standard dialog. "
    debugMessage("Open File")


def saveFile():
    " Saves a file using the standard dialog. "
    debugMessage("Save File")


def saveAsFile():
    " Saves an existing file as a new format using the standard dialog. "
    debugMessage("Save file as...")


def exitProgram():
    r"""
    Instead of closing using exit() this allows us to chain any checks,
    like the confirmation of the close, to the action.
    """
    debugMessage("Closing Embroidermodder 2.0.")
    exit()


def cutObject():
    r"""
    Cut acts on objects selected before the action is called.
    
    If no objects are selected an error is returned.
    """
    debugMessage("cut")


def copyObject():
    r"""
    Copy acts on objects selected before the action is called.
    
    If no objects are selected an error is returned.
    """
    debugMessage("copy")


def pasteObject():
    r"""
    """
    debugMessage("paste")


def icon16():
    r"""
    """
    debugMessage("icon16()")


def icon24():
    r"""
    """
    debugMessage("icon24()")


def icon32():
    r"""
    """
    debugMessage("icon32()")


def icon48():
    r"""
    """
    debugMessage("icon48()")


def icon64():
    r"""
    """
    debugMessage("icon64()")


def icon128():
    r"""
    """
    debugMessage("icon128()")


def mainPrint():
    r"""
    """
    debugMessage("print()")


def mainRedo():
    r"""
    """
    debugMessage("copy()")
    #View* gview = _mainWin->activeView()
    #if (gview) {
    #    gview->copy()


def mainUndo():
    r"""
    """
    debugMessage("main_paste()")
    #View* gview = _mainWin->activeView()
    #if gview:
    #    gview->paste()


def tipOfTheDay():
    r"""
    """
    debugMessage("main_paste()")


def changelog():
    r"""
    """
    debugMessage("changelog()")

    # display in a custom widget instead
    #
    # QUrl changelogURL("help/changelog.html")
    # QDesktopServices::openUrl(changelogURL)


def showAllLayers():
    r"""
    """
    debugMessage("showAllLayers()")


def freezeAllLayers():
    r"""
    """
    debugMessage("freezeAllLayers()")


def thawAllLayers():
    r"""
    """
    debugMessage("thawAllLayers()")


def lockAllLayers():
    r"""
    """
    debugMessage("lockAllLayers()")


def windowNext():
    r"""
    """
    debugMessage("windowNext()")


def windowPrevious():
    r"""
    """
    debugMessage("windowPrevious()")


def textItalic():
    r"""
    """
    settings["text_style_italic"] = not settings["text_style_italic"]


def textBold():
    r"""
    """
    settings["text_style_bold"] = not settings["text_style_bold"]


def textStrikeout():
    r"""
    """
    settings.text_style.strikeout = not settings.text_style.strikeout


def textUnderline():
    r"""
    """
    settings.text_style.underline = not settings.text_style.underline


def textOverline():
    r"""
    """
    settings.text_style.overline = not settings.text_style.overline


def layerManager():
    r"""
    """
    debugMessage("layerManager()")
    debugMessage("Implement layerManager.")
    #LayerManager layman( _mainWin,  _mainWin)
    #layman.exec()


def layerPrevious():
    r"""
    """
    debugMessage("layerPrevious()")
    debugMessage("Implement layerPrevious.")


def zoomRealtime():
    r"""
    """
    debugMessage("zoomRealtime()")
    debugMessage("Implement zoomRealtime.")


def zoomPrevious():
    r"""
    """
    debugMessage("zoomPrevious()")
    debugMessage("Implement zoomPrevious.")


def zoomWindow():
    r"""
    """
    debugMessage("zoomWindow()")
    #gview = _mainWin->activeView()
    #if gview:
    #    gview->zoomWindow()


def zoomDynamic():
    r"""
    """
    debugMessage("zoomDynamic()")
    debugMessage("Implement zoomDynamic.")


def zoomScale():
    r"""
    """
    debugMessage("zoomScale()")
    debugMessage("Implement zoomScale.")


def zoomCenter():
    r"""
    """
    debugMessage("zoomCenter()")
    debugMessage("Implement zoomCenter.")


def zoomIn():
    r"""
    """
    debugMessage("zoomIn()")


def zoomOut():
    ""
    debugMessage("zoomOut()")


def zoomSelected():
    ""
    debugMessage("zoomSelected()")


def zoomAll():
    ""
    debugMessage("zoomAll()")
    debugMessage("Implement zoomAll.")


def zoomExtents():
    ""
    debugMessage("zoomExtents()")


def panrealtime():
    ""
    debugMessage("panrealtime()")


def panpoint():
    ""
    debugMessage("panpoint()")


def panLeft():
    ""
    debugMessage("panLeft()")


def panRight():
    ""
    debugMessage("panRight()")


def panUp():
    ""
    debugMessage("panUp()")


def panDown():
    ""
    debugMessage("panDown()")


def dayVision():
    r""" TODO: Make day vision color settings.
    gview = _mainWin->activeView()
    if gview:
        gview->setBackgroundColor("#FFFFFF")
        gview->setCrossHairColor("#000000")
        gview->setGridColor("#000000")
        """
    debugMessage("day_vision()")


def nightVision():
    r""" TODO: Make night vision color settings.
    gview = _mainWin->activeView()
    if gview:
        gview->setBackgroundColor("#000000")
        gview->setCrossHairColor("#FFFFFF")
        gview->setGridColor("#FFFFFF")
        """
    debugMessage("night_vision()")


def whatsthisContextHelp():
    r"""
    """
    debugMessage("whatsthisContextHelp()")


def makeLayerCurrent():
    r"""
    """
    debugMessage("makeLayerCurrent()")


def layerSelector():
    r"""
    """
    debugMessage("main_exit()")


def mainHelp():
    r"""
    """
    debugMessage("main_help")


def settingsDialog():
    r"""
    """
    debugMessage("settings_dialog()")


def designDetails():
    r"""
    """
    debugMessage("main_exit()")


def mainCut():
    r"""
    """
    debugMessage("cut()")
    
    #gview = _mainWin->activeView()
    #if gview:
    #    gview->cut()


def mainCopy():
    r"""
    """
    debugMessage("copy()")
    #gview = _mainWin->activeView()
    #if gview:
    #    gview->copy()


def mainPaste():
    r"""
    """
    debugMessage("main_paste()")
    #gview = _mainWin->activeView()
    #if gview:
    #    gview->paste()


def showAllLayers():
    r"""
    """
    debugMessage("main_exit()")


def freezeAllLayers():
    r"""
    """
    debugMessage("freezeAllLayers()")


def thawAllLayers():
    r"""
    """
    debugMessage("main_exit()")


def lockAllLayers():
    r"""
    """
    debugMessage("main_exit()")


def unlockAllLayers():
    r"""
    """
    debugMessage("main_exit()")


def hideAllLayers():
    r"""
    """
    debugMessage("main_exit()")


def lineWeightSelector():
    r"""
    """
    debugMessage("main_exit()")


def lineTypeSelector():
    r"""
    """
    debugMessage("main_exit()")


def colorSelector():
    r"""
    """
    debugMessage("main_exit()")


def windowClose():
    r"""
    """
    debugMessage("main_exit()")


def windowTile():
    r"""
    """
    debugMessage("main_exit()")


def windowCloseAll():
    r"""
    """
    debugMessage("windowCloseAll()")


def windowCascade():
    r"""
    """
    debugMessage("windowCascade()")


def windowNext():
    r"""
    """
    debugMessage("main_exit()")


def windowPrevious():
    r"""
    """
    debugMessage("main_exit()")


def textItalic():
    r"""
    """
    debugMessage("main_exit()")

    settings.text_style.italic = not settings.text_style.italic


def textBold():
    r"""
    """
    debugMessage("main_exit()")

    settings.text_style.bold = not settings.text_style.bold


def textStrikeout():
    r"""
    """
    debugMessage("main_exit()")

    settings.text_style.strikeout = not settings.text_style.strikeout


def textUnderline():
    r"""
    """
    debugMessage("main_exit()")

    settings.text_style.underline = not settings.text_style.underline


def textOverline():
    r"""
    """
    debugMessage("main_exit()")

    settings.text_style.overline = not settings.text_style.overline


def makeLayerActive():
    r"""
    """
    debugMessage("makeLayerActive()")
    debugMessage("Implement makeLayerActive.")


def layerManager():
    r"""
    """
    debugMessage("layerManager()")
    debugMessage("Implement layerManager.")
    #/*LayerManager layman( _mainWin,  _mainWin)
    #layman.exec()
    #*/


def layerPrevious():
    r"""
    """
    debugMessage("layerPrevious()")
    debugMessage("Implement layerPrevious.")


def zoomRealtime():
    r"""
    """
    debugMessage("zoomRealtime()")
    debugMessage("Implement zoomRealtime.")


def zoomPrevious():
    r"""
    """
    debugMessage("zoomPrevious()")
    debugMessage("Implement zoomPrevious.")


def zoomWindow():
    r"""
    """
    debugMessage("zoomWindow()")
    #/*View* gview = _mainWin->activeView()
    #if (gview) 
    #    gview->zoomWindow()


def zoomDynamic():
    r"""
    """
    debugMessage("zoomDynamic()")
    debugMessage("Implement zoomDynamic.")


def zoomScale():
    r"""
    """
    debugMessage("zoomScale()")
    debugMessage("Implement zoomScale.")


def zoomCenter():
    r"""
    """
    debugMessage("zoomCenter()")
    debugMessage("Implement zoomCenter.")


def zoomIn():
    r"""
    """
    debugMessage("zoomIn()")


def zoomOut():
    r"""
    """
    debugMessage("zoomOut()")


def zoomSelected():
    r"""
    """
    debugMessage("zoomSelected()")


def zoomAll():
    r"""
    """
    debugMessage("zoomAll()")
    debugMessage("Implement zoomAll.")


def zoomExtents():
    r"""
    """
    debugMessage("zoomExtents()")


def panrealtime():
    r"""
    """
    debugMessage("panrealtime()")


def panpoint():
    r"""
    """
    debugMessage("panpoint()")


def panLeft():
    r"""
    """
    debugMessage("panLeft()")


def panRight():
    r"""
    """
    debugMessage("panRight()")


def panUp():
    r"""
    """
    debugMessage("panUp()")


def panDown():
    r"""
    """
    debugMessage("panDown()")


def settingsSnap():
    r"""
    """
    debugMessage("stub")


def settingsGrid():
    r"""
    """
    debugMessage("stub")


def settingsRuler():
    r"""
    """
    debugMessage("stub")


def settingsOrtho():
    r"""
    """
    debugMessage("stub")


def settingsPolar():
    r"""
    """
    debugMessage("stub")


def settingsQSnap():
    r"""
    """
    debugMessage("stub")


def settingsQTrack():
    r"""
    """
    debugMessage("stub")


def settingsLwt():
    r"""
    """
    debugMessage("stub")


def toggleSnap(on):
    r"""
    """
    debugMessage("StatusBarButton toggleSnap()")


def toggleGrid(on):
    r"""
    """
    debugMessage("StatusBarButton toggleGrid()")


def toggleRuler(on):
    r"""
    """
    debugMessage("StatusBarButton toggleRuler()")


def toggleOrtho(on):
    r"""
    """
    debugMessage("StatusBarButton toggleOrtho()")


def togglePolar(on):
    debugMessage("StatusBarButton togglePolar()")


def toggleQSnap(on):
    debugMessage("StatusBarButton toggleQSnap()")


def toggleQTrack(on):
    debugMessage("StatusBarButton toggleQTrack()")


def toggleLwt(on):
    debugMessage("StatusBarButton toggleLwt()")


def enableLwt():
    debugMessage("StatusBarButton enableLwt()")


def disableLwt():
    debugMessage("StatusBarButton disableLwt()")


def enableReal():
    debugMessage("StatusBarButton enableReal()")


def disableReal():
    debugMessage("StatusBarButton disableReal()")


def buildMenubar(root, menu_layout):
    r"""
    """
    debugMessage("build_menubar")
    menubar = tk.Menu(root)
    for m in menu_layout["order"]:
        debugMessage(m)
        menu_ = tk.Menu(menubar, tearoff=0)
        for item in menu_layout[m]["order"]:
            debugMessage(item)
            cmd = menu_layout[m][item]
            menu_.add_command(label=item, command=globals()[cmd])
        menubar.add_cascade(label=m, menu=menu_)
    root.config(menu=menubar)

# to stop the garbage collector stealing it
tkimg = {}

def buildButtonGrid(root, button_layout):
    r"""
    Create the toolbars in the order given by the "order" list.
    """
    global tkimg
    debugMessage("build_buttongrid")
    for toolbar in button_layout["order"]:
        debugMessage(toolbar)
        for button in button_layout[toolbar]["order"]:
            debugMessage(button)
            B = button_layout[toolbar][button]
            cmd = globals()[B["command"]]
            tkimg[button] = load_image(B["icon"])
            button_ = tk.Button(
                root,
                command=cmd,
                image=tkimg[button]
            )
            button_.grid(row=B["row"], column=B["column"])

